File: atc-brain-python/engine/geo_utils.py
"""
Geographic utility functions for coordinate conversion and distance calculations.
Uses small-step Earth approximation suitable for ranges < 100 NM.
"""

import math
from typing import Tuple
from .constants import (
    NM_PER_DEGREE_LAT,
    FT_PER_NM,
    CYYZ_LAT,
    CYYZ_LON,
)


def great_circle_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate great circle distance between two points using Haversine formula.
    
    Args:
        lat1, lon1: First point (degrees)
        lat2, lon2: Second point (degrees)
    
    Returns:
        Distance in nautical miles
    """
    # Convert to radians
    lat1_rad = math.radians(lat1)
    lat2_rad = math.radians(lat2)
    delta_lat = math.radians(lat2 - lat1)
    delta_lon = math.radians(lon2 - lon1)
    
    # Haversine formula
    a = (math.sin(delta_lat / 2) ** 2 +
         math.cos(lat1_rad) * math.cos(lat2_rad) *
         math.sin(delta_lon / 2) ** 2)
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    # Earth's radius in nautical miles (mean radius)
    earth_radius_nm = 3440.065
    distance_nm = earth_radius_nm * c
    
    return distance_nm


def flat_earth_distance(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """
    Calculate distance using flat-earth approximation (accurate for < 100 NM).
    Faster than great circle for local airspace.
    
    Args:
        lat1, lon1: First point (degrees)
        lat2, lon2: Second point (degrees)
    
    Returns:
        Distance in nautical miles
    """
    delta_lat = lat2 - lat1
    delta_lon = lon2 - lon1
    
    # Use midpoint latitude for cos correction
    mid_lat = (lat1 + lat2) / 2
    cos_lat = math.cos(math.radians(mid_lat))
    
    # Convert to nautical miles
    x_nm = delta_lon * NM_PER_DEGREE_LAT * cos_lat
    y_nm = delta_lat * NM_PER_DEGREE_LAT
    
    distance_nm = math.sqrt(x_nm ** 2 + y_nm ** 2)
    return distance_nm


def distance_to_airport(lat: float, lon: float, 
                       airport_lat: float = CYYZ_LAT,
                       airport_lon: float = CYYZ_LON) -> float:
    """
    Calculate distance from aircraft to airport.
    Uses flat-earth approximation for performance.
    
    Args:
        lat, lon: Aircraft position (degrees)
        airport_lat, airport_lon: Airport position (degrees)
    
    Returns:
        Distance in nautical miles
    """
    return flat_earth_distance(lat, lon, airport_lat, airport_lon)


def update_position(lat: float, lon: float, heading_deg: float, 
                   speed_kts: float, dt: float = 1.0) -> Tuple[float, float]:
    """
    Update aircraft position using small-step Earth approximation.
    
    Args:
        lat, lon: Current position (degrees)
        heading_deg: True heading (degrees, 0=North, 90=East)
        speed_kts: Ground speed (knots)
        dt: Time step (seconds, default 1.0)
    
    Returns:
        Tuple of (new_lat, new_lon) in degrees
    """
    # Convert speed to nautical miles per second
    speed_nm_per_sec = speed_kts / 3600.0
    
    # Distance traveled in this time step
    distance_nm = speed_nm_per_sec * dt
    
    # Convert heading to radians (0° = North, 90° = East)
    heading_rad = math.radians(heading_deg)
    
    # Calculate displacement in NM
    # Note: In aviation, heading 0° is North, 90° is East
    delta_north_nm = distance_nm * math.cos(heading_rad)
    delta_east_nm = distance_nm * math.sin(heading_rad)
    
    # Convert to lat/lon change
    delta_lat = delta_north_nm / NM_PER_DEGREE_LAT
    cos_lat = math.cos(math.radians(lat))
    delta_lon = delta_east_nm / (NM_PER_DEGREE_LAT * cos_lat)
    
    # Update position
    new_lat = lat + delta_lat
    new_lon = lon + delta_lon
    
    return new_lat, new_lon


def normalize_heading(heading: float) -> float:
    """
    Normalize heading to 0-360 range.
    
    Args:
        heading: Heading in degrees (may be outside 0-360)
    
    Returns:
        Normalized heading (0-360)
    """
    while heading < 0:
        heading += 360
    while heading >= 360:
        heading -= 360
    return heading


def heading_difference(current: float, target: float) -> float:
    """
    Calculate the shortest angular difference between two headings.
    Returns negative for left turn, positive for right turn.
    
    Args:
        current: Current heading (0-360)
        target: Target heading (0-360)
    
    Returns:
        Difference in degrees (-180 to +180)
    """
    diff = target - current
    
    # Normalize to -180 to +180
    while diff > 180:
        diff -= 360
    while diff < -180:
        diff += 360
    
    return diff


def bearing_to_point(from_lat: float, from_lon: float, 
                     to_lat: float, to_lon: float) -> float:
    """
    Calculate initial bearing from one point to another.
    
    Args:
        from_lat, from_lon: Starting point (degrees)
        to_lat, to_lon: Destination point (degrees)
    
    Returns:
        Initial bearing in degrees (0-360, 0=North)
    """
    # Convert to radians
    lat1 = math.radians(from_lat)
    lat2 = math.radians(to_lat)
    delta_lon = math.radians(to_lon - from_lon)
    
    # Calculate bearing
    x = math.sin(delta_lon) * math.cos(lat2)
    y = (math.cos(lat1) * math.sin(lat2) -
         math.sin(lat1) * math.cos(lat2) * math.cos(delta_lon))
    
    bearing_rad = math.atan2(x, y)
    bearing_deg = math.degrees(bearing_rad)
    
    return normalize_heading(bearing_deg)


def altitude_msl_to_agl(altitude_msl: float, elevation: float) -> float:
    """
    Convert altitude MSL to AGL.
    
    Args:
        altitude_msl: Altitude mean sea level (feet)
        elevation: Field elevation (feet MSL)
    
    Returns:
        Altitude above ground level (feet)
    """
    return altitude_msl - elevation


def altitude_agl_to_msl(altitude_agl: float, elevation: float) -> float:
    """
    Convert altitude AGL to MSL.
    
    Args:
        altitude_agl: Altitude above ground level (feet)
        elevation: Field elevation (feet MSL)
    
    Returns:
        Altitude mean sea level (feet)
    """
    return altitude_agl + elevation

File: atc-nextjs/src/utils/geoUtils.ts
/**
 * Geographic utility functions for coordinate conversion and distance calculations.
 * Uses the same flat-earth approximation as the Python engine for consistency.
 */

// CYYZ VOR coordinates (airport center)
export const CYYZ_LAT = 43.6777;
export const CYYZ_LON = -79.6248;

/**
 * Calculate distance from aircraft to CYYZ airport using flat-earth approximation.
 * This matches the calculation used in the Python engine for consistency.
 * 
 * @param lat - Aircraft latitude in degrees
 * @param lon - Aircraft longitude in degrees
 * @param airportLat - Airport latitude (defaults to CYYZ)
 * @param airportLon - Airport longitude (defaults to CYYZ)
 * @returns Distance in nautical miles
 */
export function calculateDistanceToAirport(
  lat: number, 
  lon: number, 
  airportLat: number = CYYZ_LAT,
  airportLon: number = CYYZ_LON
): number {
  const deltaLat = lat - airportLat;
  const deltaLon = lon - airportLon;
  
  // Convert to nautical miles using flat-earth approximation
  // 1 degree latitude ≈ 60 NM
  const x_nm = deltaLon * 60 * Math.cos((airportLat * Math.PI) / 180);
  const y_nm = deltaLat * 60;
  
  return Math.sqrt(x_nm * x_nm + y_nm * y_nm);
}

/**
 * Calculate heading from one point to another.
 * 
 * @param fromLat - Starting latitude
 * @param fromLon - Starting longitude
 * @param toLat - Destination latitude
 * @param toLon - Destination longitude
 * @returns Heading in degrees (0-360)
 */
export function calculateHeading(
  fromLat: number,
  fromLon: number,
  toLat: number,
  toLon: number
): number {
  const dLon = (toLon - fromLon) * Math.PI / 180;
  const lat1 = fromLat * Math.PI / 180;
  const lat2 = toLat * Math.PI / 180;
  
  const y = Math.sin(dLon) * Math.cos(lat2);
  const x = Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(dLon);
  
  let heading = Math.atan2(y, x) * 180 / Math.PI;
  return (heading + 360) % 360; // Normalize to 0-360
}

/**
 * Calculate sector based on distance from airport.
 * 
 * @param distance - Distance in nautical miles
 * @returns Sector name
 */
export function calculateSector(distance: number): string {
  if (distance > 30) return 'ENTRY';
  if (distance > 10) return 'ENROUTE';
  if (distance > 3) return 'APPROACH';
  return 'RUNWAY';
}

/**
 * Calculate glideslope altitude based on distance from airport.
 * Standard 3-degree glideslope approximation.
 * 
 * @param distance - Distance in nautical miles
 * @param airportElevation - Airport elevation in feet (defaults to CYYZ: 569 ft)
 * @returns Target altitude in feet
 */
export function calculateGlideslopeAltitude(
  distance: number, 
  airportElevation: number = 569
): number {
  // 3-degree glideslope: altitude = distance * 318 + airport_elevation
  // 318 ft per NM is the standard 3-degree glideslope
  return Math.max(airportElevation, distance * 318 + airportElevation);
}

/**
 * Generate realistic aircraft position for arrivals.
 * Uses proper distance calculation instead of hardcoded approximations.
 * 
 * @param flightType - Type of flight (ARRIVAL/DEPARTURE)
 * @returns Position object with calculated distance
 */
// Named entry fixes at ~55 NM from CYYZ (8 cardinal/intercardinal points)
const ENTRY_FIXES: { name: string; bearing: number }[] = [
  { name: 'BOXUM', bearing: 0 },    // N
  { name: 'DUVOS', bearing: 45 },   // NE
  { name: 'NUBER', bearing: 90 },   // E
  { name: 'KEDMA', bearing: 135 },  // SE
  { name: 'PILMU', bearing: 180 },  // S
  { name: 'VERKO', bearing: 225 },  // SW
  { name: 'IMEBA', bearing: 270 },  // W
  { name: 'RAGID', bearing: 315 },  // NW
];

function offsetPoint(lat: number, lon: number, distanceNm: number, bearingDeg: number): { lat: number; lon: number } {
  const bearingRad = (bearingDeg * Math.PI) / 180;
  return {
    lat: lat + (distanceNm / 60) * Math.cos(bearingRad),
    lon: lon + (distanceNm / (60 * Math.cos((lat * Math.PI) / 180))) * Math.sin(bearingRad),
  };
}

export function generateRealisticPosition(flightType: string = 'ARRIVAL'): {
  lat: number;
  lon: number;
  altitude_ft: number;
  heading: number;
  speed_kts: number;
  distance_to_airport_nm: number;
  entry_fix: { name: string; lat: number; lon: number } | null;
} {
  if (flightType === 'ARRIVAL') {
    // Spawn at 80-100 NM, outside the managed ENTRY sector
    const targetDistance = Math.random() * 20 + 80; // 80-100 NM
    const bearing = Math.random() * 360;
    const bearingRad = (bearing * Math.PI) / 180;

    const lat = CYYZ_LAT + (targetDistance / 60) * Math.cos(bearingRad);
    const lon = CYYZ_LON + (targetDistance / (60 * Math.cos((CYYZ_LAT * Math.PI) / 180))) * Math.sin(bearingRad);
    const actualDistance = calculateDistanceToAirport(lat, lon);

    // Assign nearest named entry fix (~55 NM on same bearing)
    const nearestFix = ENTRY_FIXES.reduce((best, fix) => {
      const diff = Math.abs(((fix.bearing - bearing + 540) % 360) - 180);
      const bestDiff = Math.abs(((best.bearing - bearing + 540) % 360) - 180);
      return diff < bestDiff ? fix : best;
    });
    const fixPos = offsetPoint(CYYZ_LAT, CYYZ_LON, 55, nearestFix.bearing);
    const entry_fix = { name: nearestFix.name, lat: fixPos.lat, lon: fixPos.lon };

    // Head toward the entry fix
    const headingToFix = calculateHeading(lat, lon, entry_fix.lat, entry_fix.lon);

    return {
      lat,
      lon,
      altitude_ft: Math.floor(Math.random() * 5000) + 28000, // FL280-FL330 (cruise)
      heading: Math.floor(headingToFix),
      speed_kts: Math.floor(Math.random() * 30) + 440, // 440-470 kts (cruise)
      distance_to_airport_nm: actualDistance,
      entry_fix,
    };
  } else {
    // Departures start at airport
    return {
      lat: CYYZ_LAT + (Math.random() - 0.5) * 0.01,
      lon: CYYZ_LON + (Math.random() - 0.5) * 0.01,
      altitude_ft: Math.floor(Math.random() * 1000) + 500,
      heading: Math.floor(Math.random() * 360),
      speed_kts: Math.floor(Math.random() * 50) + 150,
      distance_to_airport_nm: calculateDistanceToAirport(
        CYYZ_LAT + (Math.random() - 0.5) * 0.01,
        CYYZ_LON + (Math.random() - 0.5) * 0.01
      ),
      entry_fix: null,
    };
  }
}
